### ISSUE TITLE
Migrating to MobX from tree-structure in React. A working example for making Ajax calls? 

### ISSUE THREAD

**Issue by cbll** [mobxjs/mobx#983]
I have pretty much searched everywhere possible. How do I migrate from a traditional "react" component to using MobX?

Here's a link to my stackoverflow post with some sampel code: http://stackoverflow.com/questions/43869897/ajax-call-in-mobx-mobx-observer-store-is-not-available-make-sure-it-is-provid

Is it possible to ask for a working example of how you use, for example, axios in the Store, and pass(and bind) the data to a React component? I might be completely misunderstanding something here, I don't know - everyone says MobX is much easier than Redux but I feel more confused than ever :)

Only example I've found is here: https://github.com/burczu/react-mobx-ajax-example from a blog, but the code doesn't seem good(wrong invokations of functions etc..) 

Thanks for your help in advance, it's appreciated.

**Comment by jamiewinder** [comment]
This blog post might be helpful:
https://hackernoon.com/how-to-decouple-state-and-ui-a-k-a-you-dont-need-componentwillmount-cc90b787aa37

In regards to your example, this is how I've done it:

```javascript
class Store {
    @computed get parentData() {
        return fromPromise(axios.get("/api/parent"));
    };
}
```
`fromPromise` is a utility provided by the mobx-utils library (see [here](https://github.com/mobxjs/mobx-utils#frompromise)). 

Because this is a `@computed` property, it won't be resolved until it is asked for (so it works 'on demand' - at the point your component needs to access it, it'll be loaded) and it'll be memoized unless something else it observes changes. 

Note that using `fromPromise` will wrap your response (and the loading of it) into another object, so you can't use it in quite the same (you'll need to check if it's resolved before accessing the value), but most of the time this is what you want to do anyway in order to handle errors and show loading.

Hope this helps.

**Comment by cbll** [comment]
That makes sense. So if you have a React component using the value, how do you bind it to state?

I get that you use the` @observer` decorator in front of the class.. But how do I in practice bind it to the state so it can be rendered in components? Before, I'd just use `this.setState`(Or well, with `that` since it's nested in axios) in the API call, inside the React function. But how do I bind the data to the state of the class, so it can be propagated to the child components as props? Maybe this is a wrong way to even look at it? 

Thanks a lot in advance :)

**Comment by cbll** [comment]
I fiddled a bit with it. Does this resemble something useful?

```
import { observable, action, computed} from 'mobx';
const axios = require('axios');
const fromPromise = require('mobx-utils');
```


```
class Store {

    @observable dashboarddata;
    @computed get panelData() {
        return fromPromise(
            axios.get("/api/dashboard/").then(function(response){
                console.info("Received info ");
                dashboarddata = response.data;
            }).catch(function(error, status){
                console.error("Error getting info: "+ status+ " "+ error.toString());
            })
        );
    };
}

export default Store;
```

Or is it completely off?

**Comment by jamiewinder** [comment]
Here's a working example which simulates an AJAX call:
https://codesandbox.io/s/AD8Lv3JNp

Look at the Multiplier component to see how to use the `fromPromise` result. There is actually more to it than this - there is a `state` property so you can tell whether the promise is still pending, has fulfilled, or rejected, and a `value` that'll give you the resolved value or error. However, I find myself using the `case` utility (as shown in the example) to branch my code based on the state. 

Note that I don't need to set the result anywhere after my promise resolves; the value is picked out of the result (as you'll do with your AJAX response - you'll pick out the response body) and that becomes the resolved value of the promise, and therefore the `value` of the `fromPromise` result once it resolves.

BTW, the reason (or a reason) your example isn't working is that you're importing `fromPromise` incorrectly - do it using 
```
import { fromPromise } from 'mobx-utils';
```

Not demonstated in my example, but if you were to update either the `lhs` or `rhs` observable properties in the store, the `asyncMultiply` would automatically invalidate, and cause it to load again.

This can be really handy for parameterizing your web request; for example, having a `filter` property and a `filteredItems` property which loads items from a service which matches the filter - if the filter is changed, the items would reload.

Hope this is helping. Stick with it - this library is a joy to work with once it 'clicks' :)

**Comment by cbll** [comment]
I can see your example working in, well, a simple example like that. But in a real application where the Ajax call might be invoked elsewhere than the main containing component.. I don't really "get" how you bind the data to  `this.state` of the given React class. Since it's nested.

Do you pass the store as props to the child component? And still do the `@observer` decorator? It just seems like a huge hassle with little benefit, but again, I might not "get it" yet.

**Comment by jamiewinder** [comment]
You don't need to bind anything to anything. That's the great thing about MobX - it's all automatic. 

The fact that I've wrapped my component in `observer` means that the render function becomes reactive (i.e. will re-run if anything observable changes). My render function is accessing some observable properties (including `asyncMultiply`) so if these change, the component will automatically re-render. 

Since using MobX, I don't find myself using React's concept of 'state' at all; I keep all my application state separate from my components, and provide access to this through props. If there is any component-specific (i.e. internal) state then I use MobX-observable properties on the component class. 

I won't be able to provide a much more advanced example than this, but I hope you stick with it. It does shake up some of the things you learn when using React's state as your store, but it ultimately leads to a clean and logical separation of view and state which makes things incredibly easy to get on with.

**Comment by cbll** [comment]
So say I do an Ajax call and receive some data back. I want to pass a five different objects into five different children to the component. How would you do that? Pass props from the store? Then you're passing props as props... Or something along those lines. I just don't see how it would work in practice. The idea sounds good, though.

**Comment by jamiewinder** [comment]
Here's another (convoluted!) example which might answer your question:
https://codesandbox.io/s/76ERpypZr

Don't think of your AJAX call as such, think of it as a form of asynchronous data because that's how it appears to your components - they shouldn't care where it comes from, whether it's there already and ready to be used, or whether them accessing it will trigger a web request. They just need to be able to handle the case where it's resolved, hasn't yet resolved, or has rejected. 

I also don't think the concept of a 'store' has much meaning when it comes to MobX - it certainly does in Redux so it's all based around a single central data repository, but with MobX it's just observable data which might be the application state root (what I refer to as a store) or a branch or leaf of it. It doesn't make any difference - you just do the usual React thing of passing the relevant data along. 

In my new example, my `Calculator` component accessing `asyncResult` and (when it resolves) passes the relevant nested data to the `Result` components. So some access the store, some access the properties of the store, some access the properties of the properties, etc!

**Comment by cbll** [comment]
I'm still very unsure of how it translates to the React component lifecycles, especially for children of the parent. Right now, I manage to console.log the individual objects that are fetched from the axios.get, however they won't properly pass to the child components even though they print to the console from the parent..

So what confuses me is this: Apparently the ajax call is successful(since I can console.log the entire object and nested objects in the render method of Parent component), but they for some reason won't propagate to the children?

**Comment by cbll** [comment]
Earlier you wrote _Note that using fromPromise will wrap your response (and the loading of it) into another object, so you can't use it in quite the same (you'll need to check if it's resolved before accessing the value), but most of the time this is what you want to do anyway in order to handle errors and show loading._

Can you elaborate?

The data is, from what I can see, fetched correctly. I can access it in each child with no issues... But I get a bunch of errors and i can't render the page.

What exactly did you mean with the above? How do I check if it's resolved?

Previously I'd just check if it was true/false... As in:

```
if(!this.props.data){
//    return(<div>Nothing is here</div>
}
```

And likewise;

```
if(this.props.data){
// render the actual content
}
```

That doesn't seem to work now?

**Comment by jamiewinder** [comment]
You'll find the 'shape' of the `fromPromise` result here:
https://github.com/mobxjs/mobx-utils#frompromise

Basically, you have `value` which will eventually contain your value if it resolved, *or* the error if it rejected (I don't really like this behaviour myself, but there it is). You also have a `state` which will be one of 'pending' ("I'm still waiting for the value to be set"), 'fulfilled' ("I've now got the value"), or 'rejected' ("The promise rejected"). 

Your code for handling this might look like:
```javascript
if (myAsyncValue.state === 'fulfilled') {
    return <div>Resolved to {String(myAsyncValue.value)}</div>
} else if (myAsyncValue.state === 'rejected') {
    return <div>Rejected with {String(myAsyncValue.value)}</div>
} else { // 'pending'
    return <div>Loading...</div>
}
```
This works nicely with React and MobX because 'value' and 'state' are observable properties, so as with any observable / observer, the logic will rerun if anything changes.

The `fromPromise` result also has a very handy `case` property function which can simplify the above to:
```javascript
return myAsyncValue.case({
    fulfilled: (value) => <div>Resolved to {String(value)}</div>,
    rejected: (reason) => <div>Rejected with {String(reason)}</div>,
    pending: () => <div>Loading...</div>
});
```
Which is also the form I used in my examples above.

Basically, when working with `fromPromise`, you need to check the state before you access the value, because it might be unfulfilled, or there may have been a problem acquiring it.

**Comment by cbll** [comment]
Oh.. ok. And this is done.. In the React class where it's invoked, or the store itself?

How do I invoke the `@computed get panelData()` in my react class?

**Comment by cbll** [comment]
So say my store is as such.... And I can correctly invoke the function which then fetches data from the API. It works. But somehow it doesn't? What do I change?

`Stores.js `:

```
class Store {

    @observable dashboarddata = {};
    
       loadDashboard = () => {
        let that = this;
        axios.get('/api/dashboard/')
            .then(function (response) {
                that.dashboarddata = response.data;
            })
            .catch(function (error) {
                console.log(error);
            });
    };
}
```

1. I can console.log the **dashboarddata** just fine in the React component `<Parent 1/>`.
2. I can console.log any object inside **dashboarddata** to Children of the  `<Parent 1/>` component.

But somehow the above Axios call will catch the error every time because the children won't take the props anyhow(but they can console.log it?!)

It doesn't make any sense to me unfortunately..

**Comment by jamiewinder** [comment]
Just access it like any other property. For example, on [here](https://codesandbox.io/s/76ERpypZr), you can see that I pass the 'store' to the component as a prop on line 63, read the store from the props on line 46, read the [observable] `lhs`, `rhs`, and `asyncResult` properties on line 47 (which incidentally is where MobX - `observer` in particular - says 'ok, if this property changes I need to re-render). 

In other words, your store is just a data object which happens to be observable (by MobX) and your components are just React views which happen to be observers (i.e. reactive, by MobX-React).

**Comment by cbll** [comment]
It doesn't appear so; If I try to access paneldata the same way, the react component will give an error.. `Uncaught TypeError: _this.props.store.panelData.bind is not a function`

**Comment by jamiewinder** [comment]
If `panelData` is a return value of `fromPromise`, then the error message is correct; the [fromPromise return value](https://github.com/mobxjs/mobx-utils#frompromise) doesn't have a `bind` property. It has a `value`, `state`, `promise`, and `case` property. In which case you're using it incorrectly. 

Can you post a code snippet with how `panelData` is derived, and how you're trying to use it in a component? A minimal reproduction, like I produced above in my examples, would be ideal.

**Comment by cbll** [comment]
Here's a rough draft of my code, it's far more complex in reality with a lot of other modules; I'm migrating the Ajax call in, let's say, Parent1 for a start. That's what I'm attempting to achieve.

https://codesandbox.io/s/Q0OnWpgLG

**Comment by cbll** [comment]
Basically, the first thing I see in the console is `Received panel info Wed May 10 2017 16:08:07 GMT+0200 (CEST)` and console.logs from the children being rendered.. Yet still I get this error: `Error getting dashboard info: TypeError: Cannot read property 'data' of null`

**Comment by jamiewinder** [comment]
I can't get it to run because of something in axios (and it wouldn't load anyway), but this is what I'm getting at:
https://codesandbox.io/s/WnKykGlOg

Lots of comments on there to walk through what's happening.

I think you've missed a lot of what I was saying above regarding how `fromPromise` works; you're still doing the assignment to another property when the web request completes, you're not using it within a `@computed`, etc. It might be worth revisiting my examples and examining them more closely.

**Comment by cbll** [comment]
I really appreciate the help! I'm going through the thread now as well as your examples in detail.

**Comment by cbll** [comment]
I got it working with your posted solution, the issues were with some of my components(the data is propagated too fast now compared to before... So had some issues!)

Thanks again.

My last issue is, however.... How do we invoke the loading of data with a fixed interval(every minute) as before? I'm not really certain what the best approach should be?

**Comment by jamiewinder** [comment]
I'm not sure about this either. I have suggested things like an expiry on `computed` in the past (see [here](https://github.com/mobxjs/mobx/issues/874) if you're interested).

It's a bit hacky, but you can make it work using the following:

```javascript
import { computed } from 'mobx';
import { now } from 'mobx-utils';

class Store {
  @computed get dashboard() {
    now(60000); // How long to wait before expiring the data
    return (...);
  }
}
```
The reason this works is because the `now` utility allows you to observe the current time at the given interval (1 minute in this case). Your computed ends up observing this, and so despite not actually doing anything with the value, it'll still cause the computed to re-run every time it changes.

Perhaps someone can suggest another way - I'd be interested too!

**Comment by cbll** [comment]
I can create a new issue to discuss it, if you think it's a good idea - we might be a tad out of track here compared to the original issue :+1:

**Comment by jamiewinder** [comment]
Yes, that's probably wise. I think other users might be daunted by the length of this issue to discuss it now!

**Comment by cbll** [comment]
Hey man. I started my code today, and all of a sudden, the aforementioned dashboard in Stores.js throws this error... What could it be?

```
vendor.bundle.js:172054 Uncaught TypeError: promise.then is not a function
    at new PromiseBasedObservable (vendor.bundle.js:172054)
    at fromPromise (vendor.bundle.js:172141)
```

**Comment by cbll** [comment]
Thanks for all the help so far!

There's a pretty big issue now, though. With your method of doing the now(60000), the React component will go to "pending" for just ~1 second and then show the page again while the call is being made. That's not so good for the user obviously :-/ Do you know what we can do?

**Comment by cbll** [comment]
Oh dear... It's even running this ajax call with a fixed interval even when the component isn't mounted.. And tries to re-render it. I'm pulling the plug on this, it's a huge mess..

**Comment by jamiewinder** [comment]
I think you're going to have a bit of an upward struggle because you're back-porting MobX into your project. I tried to do the same with a major project I work on, this time moving away from Redux, and found it very time consuming and frustrating because I was having to learn as I went. 

The computed shouldn't be re-running if nothing is observing it, as this shows:
https://codesandbox.io/s/jEkKQnJW

(Open the console - It'll log for 10 seconds then stop once the autorun is disposer [equivalent to a component being unmounted]). 

As for your 1 second pending, it's because the async value - your dashboard - is being replaced with a pending value when it invalidates. To avoid this, you'd need to revert back to something like your previous code where you're setting an observable property when the load completes. However you may still find the fromPromise value useful to know when the dashboard data is in transition.

**Comment by cbll** [comment]
My example from yesterday makes a bit more sense to me mentally(https://codesandbox.io/s/Q0OnWpgLG) ... 

Is that how I bind and use dashboarddata as an observable?

**Comment by mweststrate** [comment]
@cbll gotten any further on this? If you are stick; I would try to make your solution really simple first; just some observable files that are updated by promises (skip `fromPromise` from now; using plain promises first for a while might make more clear why they are so useful). An async fetch should be as simple as: https://jsfiddle.net/mweststrate/a8az1nnj/3/ (rendering the data in children should be no problem, as long as they are marked with `@observable` as well)

**Comment by cbll** [comment]
Hi,
I basically re-did my previous implementation, just calling the `axios.get` in componentDidMount, and aftewards using setInterval to call it every minute. Only exception being the promise is in the store, rather than the react class(which is cool though, since I can pass the data to more components :))

**Comment by mweststrate** [comment]
Great to hear! That means I can close the issue ;-) ?

**Comment by cbll** [comment]
Indeed, thanks for the help on here :) Will definitely use MobX next time I make a new react project(rather than porting a tree-structure again.. because that was painful! :P )
